x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java

```diff
@@ -500,48 +355,167 @@ private GetAliasesRequest createGetAliasesRequest(FieldCapabilitiesResponse resp
+            boolean isMergedMapping) {
+        Map<String, QlFieldCapabilities> map = new HashMap<>();
+        Map<String, Set<String>> aliasToIndices = new HashMap<>(); // map with aliases and their list of indices
+        
+        if (indexToAliases != null) {
+            Iterator<ObjectObjectCursor<String, List<AliasMetadata>>> iter = indexToAliases.iterator();
+            while (iter.hasNext()) {
+                ObjectObjectCursor<String, List<AliasMetadata>> index = iter.next();
+                for (AliasMetadata aliasMetadata : index.value) {
+                    String aliasName = aliasMetadata.alias();
+                    aliasToIndices.putIfAbsent(aliasName, new HashSet<>());
+                    aliasToIndices.get(aliasName).add(index.key);
+                }
+            }
+        } else if (isMergedMapping) {
+            // single virtual alias named "*" to be used for a merged mappings
+            aliasToIndices.put("*", new HashSet<>(Arrays.asList(indexNames)));
+        }
+
+        // create a virtual set of fields capabilities for each alias
+        for (String alias : aliasToIndices.keySet()) {
+            Set<String> indices = aliasToIndices.get(alias);
+            QlFieldCapabilities qlFieldCaps = new QlFieldCapabilities(indices.toArray(new String[indices.size()]));
+            Map<String, Map<String, QlFieldCapability>> fields = new HashMap<>();
+            
+            // check each field
+            for (Entry<String, Map<String, FieldCapabilities>> entry : fieldCaps.entrySet()) {
+                String fieldName = entry.getKey();
+                
+                // ignore size added by the mapper plugin
+                if (FIELD_NAMES_BLACKLIST.contains(fieldName)) {
+                    continue;
+                }
+                Map<String, QlFieldCapability> qlFieldTypes = new HashMap<>();
+
+                // check each type
+                for (Entry<String, FieldCapabilities> typeEntry : entry.getValue().entrySet()) {
+                    String type = typeEntry.getKey();
+                    FieldCapabilities f = typeEntry.getValue();
+                    // Skip internal fields (name starting with underscore and its type reported by field_caps starts
+                    // with underscore as well). A meta field named "_version", for example, has the type named "_version".
+                    if (type.startsWith("_") && f.getType().startsWith("_")) {
+                        continue;
+                    }
+                    
+                    QlFieldCapability qlFieldCap = new QlFieldCapability(type, f.getType(), f.isSearchable(), f.isAggregatable());
+                    boolean isValidType = false;
+
+                    // in each list of indices, keep only those belonging to the current alias
+                    isValidType = maybeSetIndices(f.indices(), indices, qlFieldCap::setIndices) || isValidType;
```

Comment:
Can you explain what this does? If it's an OR, making the right side left (`isValidType ||`) would skip the method evaluation if `isValidType` is true which is surprising since it means that once a type is valid, it will always remain as such.
If that's not the case why go for `||`?

Reply:
`maybeSetIndices` checks the common indices between the two collections received as parameter and sets them as an array on the `QlFieldCapability` object based on the method received as the third parameter. If the common indices list is valid then the list is set and method returns `true`. Otherwise `false`.

`maybeSetIndices` needs to always be called, thus being first in the `||`.
And once one of these three calls returns `true`, the value of `isValidType` shouldn't change, thus `|| isValidType` is called last (anything `|| true` is `true`).